Trivial change arf

ezyang opened the pull request:
Trivial change arf

Summary: none

Differential Revision: https://phabricator.intern.facebook.com/D9514646

apaszke commented:
One apparent thing is that it would be much more convenient to use `std::list<Node>`, which would also give us Node lifetime management for free. From what I saw so far, we always know where in the canonical order will we place our node, so we could as well do that when it is constructed. I'll add that if you're ok with this change

apaszke commented:
Also, this commit is untested. I don't really know how to check the fuser. This is why we should start including unit tests. I can work on exposing the IR internals in Python if that's needed

zdevito commented on torch/csrc/jit/graph_fuser.cpp:
```diff
@@ -19,21 +19,24 @@ struct GraphFuser {
-          it.eraseCurrentFromParent();
+          single_select->destroy();
+          // Erasing p directly would invalidate iterator
+          --it;
```
If `it == nodes.begin()`, I believe it is undefined behavior to iterate before `begin()`

zdevito commented on torch/csrc/jit/ir.h:
```diff
@@ -234,12 +222,7 @@ struct Node {
-    this->prev() = n;
-    this->next() = next;
-    next->prev() = this;
+    insertBefore(n->next());
```
I don't think `n->next()` is valid here where `n` is at the end of the list. Previously this was valid because the list was circular.

apaszke replied:
I think it's ok, because `container.end()` points one past the last element, so in worst case it should point to it

zdevito replied:
But isn't  `n->next()` a Node*, not a `std::list::iterator`?

apaszke replied:
Yeah right. I already fixed it locally, but I'm still squashing a few other bugs

zdevito commented on torch/csrc/jit/ir.h:
```diff
@@ -553,55 +471,68 @@ friend struct NodeWithKind;
+  nodes_iter_ = graph_->nodes_.end();
+}
+
+inline void Node::destroy() {
```
Mentioned this in the comments of the fuser PR, but I'd like to keep this name because it aligns well with the LLVM name for the same thing and will make more sense once we have control flow. If you really feel strongly about `Parent` for now, then `eraseFromGraph` is close enough to the LLVM that people familiar with LLVM will get the connection.

zdevito commented:
I think we should stick with `std::list<Node*>` I mentioned in the other PR: with control flow there will be many basic blocks, each holding a nodes list, and it will be common to move nodes across blocks. So I don't want to couple our ownership model to the nodes list in anticipation of that behavior. I also would imagine there will be cases in optimizations where you will need to create a node before you know where to add it.

apaszke commented:
The main downside of `std::list<Node*>` is that you have to dereference the iterator twice to actually do anything with the node, which is pretty annoying. However, I realized that it should be able to use `std::list<Node&>` since all nodes are alive for at least as long as they're members of the list. Does that sound ok to you?

Ad LLVM naming conventions. I think the name is pretty bad for what this function does (it effectively invalidates **all** references to the node). I know that it might be more familiar to those with LLVM background, but it's complete nonsense for everyone else. Let's talk about this in today's standup.

apaszke commented:
I've updated the PR and verified that `test/test_jit.py` runs

ezyang committed:
Trivial change arf

Summary: none

Differential Revision: https://phabricator.intern.facebook.com/D9514646
